File: assets/web/js/modules/sms/smsTextInputBridge.js
/**
 * @file smsTextInputBridge.js
 * @description Minimal bridge to handle Flutter text input events for SMS functionality.
 * This file focuses ONLY on text input bridging - no other SMS features.
 * 
 * Phase 1: Text Input Event Handling
 * - Listen for flutter-text-input events
 * - Provide debugging and validation
 * - Forward to SMS Integration Manager
 */

class SmsTextInputBridge {
    constructor() {
        this.isActive = false;
        this.currentContactId = null;
        this.debugMode = true; // Enable for testing phase
        
        console.log("📝 SMS Text Input Bridge initialized");
        this.setupEventListeners();
    }

    /**
     * Setup event listeners for Flutter text input
     */
    setupEventListeners() {
        // Listen for Flutter text input events
        window.addEventListener('flutter-text-input', (event) => {
            this.handleFlutterTextInput(event);
        });

        // Also listen for any SMS input responses that might contain text
        window.addEventListener('flutter-sms-input-response', (event) => {
            this.handleSmsInputResponse(event);
        });

        console.log("📝 Text input event listeners registered");
    }

    /**
     * Handle optimized Flutter text input events
     * PERFORMANCE OPTIMIZATION: Now processes complete messages instead of character-by-character updates
     * @param {CustomEvent} event - The Flutter text input event
     */
    handleFlutterTextInput(event) {
        if (this.debugMode) {
            console.log("📝 🔍 OPTIMIZED FLUTTER TEXT INPUT EVENT:");
            console.log("📝 Event type:", event.type);
            console.log("📝 Event detail:", event.detail);
            console.log("📝 Bridge active:", this.isActive);
            console.log("📝 Current contact:", this.currentContactId);
            console.log("📝 Optimized flag:", event.detail?.optimized);
        }

        // Validate event structure
        if (!event.detail) {
            console.warn("📝 ⚠️ Text input event missing detail object");
            return;
        }

        // Extract text from event
        const text = event.detail.text || '';
        const contactId = event.detail.contactId || this.currentContactId;
        const isOptimized = event.detail.optimized || false;

        if (this.debugMode) {
            console.log(`📝 ✅ Processing ${isOptimized ? 'OPTIMIZED COMPLETE MESSAGE' : 'text input'}: "${text}" for contact: ${contactId}`);
        }

        // Forward to SMS Integration Manager if available
        if (window.smsIntegrationManager && typeof window.smsIntegrationManager.handleChannelTextInput === 'function') {
            window.smsIntegrationManager.handleChannelTextInput(text, contactId);
            console.log(`📝 ✅ ${isOptimized ? 'Complete message' : 'Text'} forwarded to SMS Integration Manager`);
        } else {
            console.warn("📝 ⚠️ SMS Integration Manager not available for text forwarding");
        }

        // Dispatch custom event for other listeners
        window.dispatchEvent(new CustomEvent('sms-text-updated', {
            detail: { text, contactId, optimized: isOptimized }
        }));
    }

    /**
     * Handle SMS input response events that might contain text
     * @param {CustomEvent} event - The SMS input response event
     */
    handleSmsInputResponse(event) {
        if (!event.detail) return;

        // Check if this response contains text content
        if (event.detail.text !== undefined) {
            if (this.debugMode) {
                console.log("📝 🔍 SMS INPUT RESPONSE WITH TEXT:");
                console.log("📝 Text:", event.detail.text);
                console.log("📝 Action:", event.detail.action);
                console.log("📝 Contact:", event.detail.contactId);
            }

            // Treat this as a text input event
            this.handleFlutterTextInput({
                type: 'flutter-text-input',
                detail: {
                    text: event.detail.text,
                    contactId: event.detail.contactId || this.currentContactId
                }
            });
        }
    }

    /**
     * Activate text input bridging for a specific contact
     * @param {string} contactId - The contact ID
     */
    activate(contactId) {
        this.isActive = true;
        this.currentContactId = contactId;
        console.log(`📝 ✅ Text input bridge activated for contact: ${contactId}`);
    }

    /**
     * Deactivate text input bridging
     */
    deactivate() {
        this.isActive = false;
        this.currentContactId = null;
        console.log("📝 ❌ Text input bridge deactivated");
    }

    /**
     * Toggle debug mode for testing
     * @param {boolean} enabled - Whether to enable debug mode
     */
    setDebugMode(enabled) {
        this.debugMode = enabled;
        console.log(`📝 Debug mode ${enabled ? 'enabled' : 'disabled'}`);
    }

    /**
     * Get current state for debugging
     * @returns {object} Current state
     */
    getState() {
        return {
            isActive: this.isActive,
            currentContactId: this.currentContactId,
            debugMode: this.debugMode
        };
    }
}

// Create global instance
window.smsTextInputBridge = new SmsTextInputBridge();

// Export for module usage
if (typeof module !== 'undefined' && module.exports) {
    module.exports = SmsTextInputBridge;
}

console.log("📝 SMS Text Input Bridge loaded and ready for testing");

File: FirstTapsMV3D_v15REF200/lib/sms_text_input_bridge.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

/// SMS Text Input Bridge for Flutter
///
/// Phase 1: Text Input Event Handling
/// This class handles TextField input and sends text events to JavaScript
///
/// Usage:
/// 1. Create an instance: final bridge = SmsTextInputBridge();
/// 2. Activate for contact: bridge.activateForContact(contactId);
/// 3. Use bridge.textController in your TextField
/// 4. Call bridge.deactivate() when done

class SmsTextInputBridge {
  static const String _jsChannelName = 'flutter-text-input';

  final TextEditingController textController = TextEditingController();
  String? currentContactId;
  bool isActive = false;
  bool debugMode = kDebugMode; // Auto-disabled in release builds

  SmsTextInputBridge() {
    _setupTextListener();
    _log("SMS Text Input Bridge initialized");
  }

  /// Setup text input listener
  void _setupTextListener() {
    textController.addListener(() {
      if (isActive && currentContactId != null) {
        final text = textController.text;
        _sendTextToJavaScript(text);
      }
    });
  }

  /// Activate text input bridging for a specific contact
  void activateForContact(String contactId) {
    currentContactId = contactId;
    isActive = true;
    textController.clear(); // Start with clean slate
    _log("Text input bridge activated for contact: $contactId");
  }

  /// Deactivate text input bridging
  void deactivate() {
    isActive = false;
    currentContactId = null;
    textController.clear();
    _log("Text input bridge deactivated");
  }

  /// Send text input to JavaScript
  void _sendTextToJavaScript(String text) {
    if (!isActive || currentContactId == null) return;

    try {
      final event = {
        'text': text,
        'contactId': currentContactId,
        'timestamp': DateTime.now().millisecondsSinceEpoch,
      };

      _log("Sending text to JS: '$text' for contact: $currentContactId");

      // Send via JavaScript channel
      _dispatchJavaScriptEvent(_jsChannelName, event);
    } catch (e) {
      _log("Error sending text to JavaScript: $e");
    }
  }

  /// Dispatch JavaScript event
  void _dispatchJavaScriptEvent(String eventName, Map<String, dynamic> data) {
    // This would be implemented based on your specific Flutter-JS bridge
    // For now, this is a placeholder that shows the intended structure

    // Example for flutter_inappwebview:
    // webViewController?.evaluateJavascript(source: '''
    //   window.dispatchEvent(new CustomEvent('$eventName', {
    //     detail: ${jsonEncode(data)}
    //   }));
    // ''');

    _log("Would dispatch JS event: $eventName with data: $data");
  }

  /// Create a TextField widget that uses this bridge
  Widget createTextField({
    String? hintText,
    TextStyle? style,
    InputDecoration? decoration,
  }) {
    return TextField(
      controller: textController,
      style: style,
      decoration:
          decoration ??
          InputDecoration(
            hintText: hintText ?? 'Type your message...',
            border: OutlineInputBorder(),
            contentPadding: EdgeInsets.symmetric(horizontal: 16, vertical: 12),
          ),
      onChanged: (text) {
        // Text changes are automatically handled by the controller listener
        _log("TextField onChanged: '$text'");
      },
      onSubmitted: (text) {
        _log("TextField onSubmitted: '$text'");
        // Optionally clear the field after submission
        // textController.clear();
      },
    );
  }

  /// Get current state for debugging
  Map<String, dynamic> getState() {
    return {
      'isActive': isActive,
      'currentContactId': currentContactId,
      'textLength': textController.text.length,
      'debugMode': debugMode,
    };
  }

  /// Toggle debug mode
  void setDebugMode(bool enabled) {
    debugMode = enabled;
    _log("Debug mode ${enabled ? 'enabled' : 'disabled'}");
  }

  /// Debug logging
  void _log(String message) {
    if (debugMode) {
      print("📝 [SmsTextInputBridge] $message");
    }
  }

  /// Dispose resources
  void dispose() {
    textController.dispose();
    _log("SMS Text Input Bridge disposed");
  }
}

/// Example usage widget for testing
class SmsTextInputTestWidget extends StatefulWidget {
  final String contactId;

  const SmsTextInputTestWidget({Key? key, required this.contactId})
    : super(key: key);

  @override
  _SmsTextInputTestWidgetState createState() => _SmsTextInputTestWidgetState();
}

class _SmsTextInputTestWidgetState extends State<SmsTextInputTestWidget> {
  late SmsTextInputBridge bridge;

  @override
  void initState() {
    super.initState();
    bridge = SmsTextInputBridge();
    bridge.activateForContact(widget.contactId);
  }

  @override
  void dispose() {
    bridge.deactivate();
    bridge.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(16),
      child: Column(
        children: [
          Text(
            'SMS Text Input Test',
            style: Theme.of(context).textTheme.headlineSmall,
          ),
          SizedBox(height: 16),
          bridge.createTextField(hintText: 'Type to test text input bridge...'),
          SizedBox(height: 16),
          ElevatedButton(
            onPressed: () {
              final state = bridge.getState();
              ScaffoldMessenger.of(
                context,
              ).showSnackBar(SnackBar(content: Text('Bridge state: $state')));
            },
            child: Text('Check Bridge State'),
          ),
        ],
      ),
    );
  }
}
